### PR TITLE
Contracts refactoring

### PR DESCRIPTION
**Pull request by indeavr**
Please carefully review. 
Changes include:
1) using SafeERC20 for IERC20
2) depositPercent per job
3) is Ownable
4) many onlyOwner management functions
5) COMPLETED_TIMEOUT_SECONDS = 7 days
6) Lots of TODO comments and questions for us to discuss. 
7) Some memory improvements

8) paymentTokens is now an array of ERC20
* jobData includes token field
* daoTreasury address included which receives all the DAO_FEEs as ERC20 tokens. 
The DaoTreasury could also be a Vault contract that automatically swaps all ERC20s to a token of our choosing. 
 
9) removed `daoEscrow` and `daoFunds` properties. Use `erc20.balanceOf(jobContractAddr)` instead 
10) disputeResolver address introduced 
11) linting fixes. 

After you review and approve we need to refactor the tests.

### CODE REVIEW

File: contracts/contracts/Job.sol [hunk 1]
```diff
@@ -73,96 +77,120 @@ contract Job {
...
+     */
+    function postJob(
+        IERC20 paymentToken,
+        uint256 bountyValue,
+        uint256 depositPct,
+        string memory jobMetaData
+    ) public onlyWhitelisted(paymentToken) requiresApproval(paymentToken, bountyValue) {
+        // TODO: add jobMetaData length check after ipfs integration is ready.
+        require(bountyValue >= MINIMUM_BOUNTY, "Minimum payment not provided");
+        require(depositPct < BASE_PERCENTAGE, "Minimum payment not provided");
```
Comment (deweller) [inline contracts/contracts/Job.sol:175]: Error should be "Deposit percent is too high"

File: contracts/contracts/Job.sol [hunk 2]
```diff
@@ -233,148 +262,233 @@ contract Job {
...
+    // TODO: Do we need any convenient getters ?
+
     /****************************
      * DAO Management Functions *
      ****************************/
 
-    function withdrawDaoFunds(address recipient, uint amount) public requiresFundManager() {
-        require(amount <= daoFunds, "Insufficient funds");
+    // TODO: Add a safety function to make sure that funds can never be locked in the contract.
+    // TODO: "initiateWithdraw" which starts a 24h timer after which "withdraw(jobId)" can be called by onlyOwner()
```
Comment (deweller) [inline contracts/contracts/Job.sol:348]: I think the dispute resolution covers this case, although it does take a fee.

File: contracts/contracts/Job.sol [hunk 3]
```diff
@@ -233,148 +262,233 @@ contract Job {
...
+
+    // TODO: what if someone sends a token by mistake to this contract ?
```
Comment (deweller) [inline contracts/contracts/Job.sol:350]: I thought about this a bit too.  By allowing arbitrary withdrawals of funds, we create a larger surface area for an attack.  If someone gets the contract private key(s) they could drain both the daoFunds and the daoEscrow if we have an emergency withdrawal function.

I think we should be conservative and just not allow arbitrary ERC20 withdrawals even by the owner.

We might want to add a function to withdraw ETH, though...

File: contracts/contracts/Job.sol [hunk 4]
```diff
@@ -233,148 +262,233 @@ contract Job {
...
 
-        daoFunds = daoFunds - amount;
-        sendFunds(recipient, amount);
+    function updatePaymentTokens(IERC20 token, bool value) external onlyOwner {
+        paymentTokens[token] = value;
+        emit PaymentTokensUpdated(token, value);
+    }
+
+    // TODO: all these functions can either be with a Timelocker or with constrained values (see setJobTimeout).
+    // TODO: So that people don't have to trust us
```
Comment (deweller) [inline contracts/contracts/Job.sol:358]: These are good.  Some should apply globally (setDefaultDeposit).  Some might apply per job (setResolutionFee).  We should discuss and decide.

I think bounds checking is good for all of them.

**Comment by deweller** [comment]
Good work.  I added a few suggestions and a few things the DAO needs to decide.

**Review by deweller — CHANGES_REQUESTED** [review]
Good work here.  I think it is a good idea to keep the contract tests in sync whenever we change or add something.

File: contracts/contracts/Job.sol [hunk 5]
```diff
@@ -2,61 +2,65 @@
...
+        address supplier;
+        address engineer;
+        IERC20 token;
+
+        // Min % of bounty that an engineer needs to deposit to start the job
+        uint256 depositPct;
+        uint256 deposit;
+        uint256 bounty;
+        uint256 startTime;
+        uint256 completedTime;
```
[the remark this answers is not in the document]
Comment (indeavr) [reply contracts/contracts/Job.sol:62]: Nah

File: contracts/contracts/Job.sol [hunk 6]
```diff
@@ -233,148 +262,305 @@ contract Job {
...
 
-    function withdrawDaoFunds(address recipient, uint amount) public requiresFundManager() {
-        require(amount <= daoFunds, "Insufficient funds");
+    // TODO: what if someone sends a token by mistake to this contract ?
+    // TODO: function withdraw
+
+    function updatePaymentTokens(IERC20 token, bool enable) external onlyOwner {
+        paymentTokens[token] = enable;
+        if (enable) {
+            tokensList.push(token);
```
[the remark this answers is not in the document]
Comment (deweller) [reply contracts/contracts/Job.sol:387]: Created #52

File: contracts/contracts/Job.sol (hunk 4 above)
Comment (deweller) [reply contracts/contracts/Job.sol:395]: Created issue #53

**Review by deweller — APPROVED** [review]
Great work.

I added additional issues for potential next steps so we can discuss and/or implement them one and a time.

Ack on merging this.

File: contracts/contracts/Job.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (indeavr) [reply contracts/contracts/Job.sol:387]: lol, Good catch !

**Review by reisr3 — APPROVED** [review]

**Pull request merged**